File: 54-spiral-matrix/54-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        
        int s = 0;
        int[][] dir = new int[][] {
            {0, 1}, {1, 0}, {0, -1}, {-1, 0}
        };
        
        int r = matrix.length;
        int c = matrix[0].length;
        int[] ep = new int[] {-1, r, -1, c};
        
        int total = r * c, i = 0, j = 0;
        while(res.size() < total) {
            res.add(matrix[i][j]);
            
            if (end(ep, i + dir[s][0], j + dir[s][1])) {
                s = (s + 1) % dir.length;
                int[] p = dir[s];
                ep[0] = Math.max(ep[0], ep[0] + p[0]);
                ep[1] = Math.min(ep[1], ep[1] + p[0]);
                ep[2] = Math.max(ep[2], ep[2] + p[1]);
                ep[3] = Math.min(ep[3], ep[3] + p[1]);
            }
            
            i += dir[s][0];
            j += dir[s][1];
        }
        
        return res;
    }
    
    boolean end(int[] ep, int i, int j) {
        return i <= ep[0] || i >= ep[1] || j <= ep[2] || j >= ep[3];
    }
}

File: 54-spiral-matrix/54-spiral-matrix.scala
import scala.collection.mutable._

object Solution {
    def spiralOrder(matrix: Array[Array[Int]]): List[Int] = {
        val r = matrix.length
        val c = matrix(0).length
        val dirs = Array(Array(0, 1), Array(1, 0), Array(0, -1), Array(-1, 0))
        var rowRange: Array[Int] = Array(-1, r)
        var colRange: Array[Int] = Array(-1, c)
        var curDir = 0
        val res: ListBuffer[Int] = ListBuffer()
        var curR = 0
        var curC = 0

        while (res.length < r * c) {
            val dir = dirs(curDir)

            while (
                curR > rowRange(0) && curR < rowRange(1) && curC > colRange(0) && curC < colRange(1)
            ) {
                res += matrix(curR)(curC)
                curR += dir(0)
                curC += dir(1)
            }

            curR -= dir(0)
            curC -= dir(1)

            curDir = (curDir + 1) % 4
            val nextDir = dirs(curDir)

            curR += nextDir(0)
            curC += nextDir(1)

            rowRange(0) = Math.max(rowRange(0), rowRange(0) + nextDir(0))
            rowRange(1) = Math.min(rowRange(1), rowRange(1) + nextDir(0))
            colRange(0) = Math.max(colRange(0), colRange(0) + nextDir(1))
            colRange(1) = Math.min(colRange(1), colRange(1) + nextDir(1))
        }

        res.toList
    }
}